src/frequenz/channels/util/_periodic_timer.py

```diff
@@ -0,0 +1,201 @@
+        `_stopped` is `False`) or was explicitly stopped (if `_stopped` is
+        `True`).
+        """
+
+        self._current_drift: timedelta | None = None
+        """The difference between `_next_msg_time` and the actual time when the timer fired.
+
+        This is calculated by `ready()` but is returned by `consume()`. If
+        `None` it means `ready()` wasn't called and `consume()` will assert.
+        `consume()` will set it back to `None` to tell `ready()` that it needs
+        to wait again.
+        """
+
+        if auto_start:
+            self.reset()
+
+    def reset(self) -> None:
+        """Reset the timer to start timing from now.
+
+        If the timer was stopped, or not started yet, it will be started.
+
+        This can only be called with a running loop, see the class
+        documentation for more details.
+        """
+        self._stopped = False
+        self._next_msg_time = asyncio.get_running_loop().time() + self._interval
+
+    def stop(self) -> None:
+        """Stop the timer.
+
+        Once `stop` has been called, all subsequent calls to `ready()` will
+        immediately return False and calls to `consume()` / `receive()` or any
+        use of the async iterator interface will raise
+        a `ReceiverStoppedError`.
+
+        You can restart the timer with `reset()`.
+        """
+        self._stopped = True
+        self._next_msg_time = None
+
+    async def ready(self) -> bool:
+        """Wait until the timer interval passed.
+
+        Once a call to `ready()` has finished, the resulting drift must be read
+        with a call to `consume()` (`receive()` or iterated over) to tell the
+        timer it should wait for the next interval.
+
+        The timer will remain ready (this method will return immediately)
+        until it is consumed.
+
+        Returns:
+            Whether the timer is still running (if it was `stop()`ed, it will
+                return `False`.
+
+        Raises:
+            RuntimeError: if it was called without a running loop.
+        """
+        # if there are messages waiting to be consumed, return immediately.
+        if self._current_drift is not None:
+            return True
+
+        # If a stop was explicitly requested, we bail out.
+        if self._stopped:
+            return False
+
+        # If there is no stop requested but we don't have a time for the next
+        # message, then we reset() (so we start the timer as of now).
+        if self._next_msg_time is None:
+            self.reset()
+        assert self._next_msg_time is not None, "This should be assigned by reset()"
+
+        now = asyncio.get_running_loop().time()
+        diff = self._next_msg_time - now
+        if diff > 0:
+            await asyncio.sleep(diff)
+            now = asyncio.get_running_loop().time()
+        self._current_drift = timedelta(seconds=now - self._next_msg_time)
+        self._next_msg_time = self._next_msg_time + self._interval
```

Comment:
if drift is greater than the interval,  do we want to directly skip over to the next tick that comes after the current time or do we make a tick for each missed interval?  I feel that we should just skip over the missed ones because we're just reporting clock drifts anyway, and not timestamps.

Reply:
Good question. I am emitting the tick because this was intended to be used by the resampler, and we want to keep resampling for all the periods even if they are late. Maybe we can make it a config option.

Reply:
I was thinking that since from now we only have a use case for the tick-always version, I would rather keep it like this for now and add the other option later, if needed.